man/man1/mininode.1
.TH MININODE 1 2016 MiniNode.js MiniNode.js

.\ This is a man page comment.
.\ Man page syntax (actually troff syntax) is somewhat obscure, but the
.\ important part is is that .<letter> specifies <letter>'s syntax for that
.\ line, and \f<letter> specifies it for the characters that follow.
.\ See http://liw.fi/manpages/ for more info.

.SH NAME

mininode \- A pure C implementation of the Node.js API


.SH SYNOPSIS

.B node
[\fIoptions\fR]
\fIscript.js\fR
.br

Execute without arguments to start the REPL.


.SH DESCRIPTION

MiniNode.js is an implementation of the Node.js 6.2.2 API using the Duktape
Javascript interpreter and a pure C implementation of the Node standard library.
It is primarily focused on the rapid development of embedded clients.


.SH OPTIONS

.TP
.BR \-v ", " \-\-version
Print node's version.

.TP
.BR \-h ", " \-\-help
Print node command line options.
The output of this option is less detailed than this document.

.TP
.BR \-e ", " \-\-eval " " \fR"\fIscript\fR"
Evaluate the following argument as JavaScript.

.TP
.BR \-p ", " \-\-print " " \fR"\fIscript\fR"
Identical to \fB-e\fR but prints the result.

.TP
.BR \-c ", " \-\-check
Syntax check the script without executing.

.TP
.BR \-i ", " \-\-interactive
Opens the REPL even if stdin does not appear to be a terminal.

.TP
.BR \-r ", " \-\-require " " \fImodule\fR
Preload the specified module at startup. Follows `require()`'s module resolution
rules. \fImodule\fR may be either a path to a file, or a node module name.

.TP
.BR \-\-no\-deprecation
Silence deprecation warnings.

.TP
.BR \-\-trace\-deprecation
Print stack traces for deprecations.

.TP
.BR \-\-throw\-deprecation
Throw errors for deprecations.

.TP
.BR \-\-trace\-sync\-io
Prints a stack trace whenever synchronous I/O is detected after the first turn
of the event loop.

.TP
.BR \-\-track\-heap-objects
Track heap object allocations for heap snapshots.

.TP
.BR \-\-zero\-fill\-buffers
Automatically zero-fills all newly allocated Buffer and SlowBuffer instances.

.TP
.BR \-\-prof\-process
Process v8 profiler output generated using the v8 option \fB\-\-prof\fR

.TP
.BR \-\-v8\-options
Print v8 command line options.

.TP
.BR \-\-v8\-pool\-size =\fInum\fR
Set v8's thread pool size which will be used to allocate background jobs.
If set to 0 then v8 will choose an appropriate size of the thread pool based
on the number of online processors. If the value provided is larger than v8's
max then the largest value will be chosen.

.TP
.BR \-\-tls\-cipher\-list =\fIlist\fR
Specify an alternative default TLS cipher list. (Requires Node.js to be built with crypto support. (Default))

.TP
.BR \-\-enable\-fips
Enable FIPS-compliant crypto at startup. (Requires Node.js to be built with \fB./configure \-\-openssl\-fips\fR)

.TP
.BR \-\-force\-fips
Force FIPS-compliant crypto on startup. (Cannot be disabled from script code.) (Same requirements as \fB\-\-enable\-fips\fR)

.TP
.BR \-\-icu\-data\-dir =\fIfile\fR
Specify ICU data load path. (overrides \fBNODE_ICU_DATA\fR)

.SH ENVIRONMENT VARIABLES

.TP
.BR NODE_DEBUG =\fImodule\fR[,\fI...\fR]
\',\'\-separated list of core modules that should print debug information.

.TP
.BR NODE_PATH =\fIpath\fR[:\fI...\fR]
\':\'\-separated list of directories prefixed to the module search path.

.TP
.BR NODE_DISABLE_COLORS=1
When set to 1 colors will not be used in the REPL.

.TP
.BR NODE_ICU_DATA =\fIfile\fR
Data path for ICU (Intl object) data. Will extend linked-in data when compiled
with small\-icu support.

.TP
.BR NODE_REPL_HISTORY =\fIfile\fR
Path to the file used to store the persistent REPL history. The default path
is ~/.node_repl_history, which is overridden by this variable. Setting the
value to an empty string ("" or " ") disables persistent REPL history.


.SH RESOURCES AND DOCUMENTATION

Website: \fBhttps://nodejs.org/\fR

Documentation: \fBhttps://nodejs.org/api/\fR

GitHub repository & Issue Tracker: \fBhttps://github.com/nodejs/node\fR

Mailing list: \fBhttp://groups.google.com/group/nodejs\fR

IRC (general questions): \fBchat.freenode.net #node.js\fR

IRC (node core development): \fBchat.freenode.net #node-dev\fR


src/mininode.c
/**
 * mininode - Node.js for Embedded Systems.
 *
 * Copyright (c) 2016 Alex Caudill. All rights reserved.
 *
 * See LICENSE.txt for the MIT license.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <string.h>
#include <limits.h>
#include <getopt.h>
#include "uv.h"
#include "duktape.h"
#include "mininode.h"
#include "modules.h"
#include "builtin_hash.h"

/* Accepted flags for getopt() */
#define OPTSTRING "chipvz"
/* Flag set by -p */
static int print_flag = 0;
/* Flag set by -c */
static int check_flag = 0;
/* Flag set by -i */
static int interactive_flag = 0;
/* Flag set by --z */
static int zero_fill_flag = 0;

void
mn_short_help() {
	printf("Usage: mininode [options] script.js\n");
}

/*
 * Several of these utility functions are based on the Duktape examples.
 * The example code will gradually disappear, but I'm making a note!
 */
void
mn_sighandler(int x) {
	/*
	 * TODO: Signals should be passed upstack to the
	 * Javascript runtime. For now, just print them.
	 */
	fprintf(stderr, "Got signal %d\n", x);
	fflush(stderr);
}

void
mn_set_sigint_handler(void) {
	(void)signal(SIGINT, mn_sighandler);
	(void)signal(SIGPIPE, SIG_IGN);  /* avoid SIGPIPE killing process */
}

void
mn_fatal_handler(const char *msg) {
	fprintf(stderr, "*** FATAL ERROR: %s\n", msg ? msg : "no message");
	fprintf(stderr, "Causing intentional segfault...\n");
	fflush(stderr);
	*((volatile unsigned int *) 0) = (unsigned int) 0xdeadbeefUL;
	abort();
}

int
mn_get_stack_raw(duk_context *ctx) {
	if (!duk_is_object(ctx, -1)) {
		return 1;
	}

	if (!duk_has_prop_string(ctx, -1, "stack")) {
		return 1;
	}

	if (!duk_is_error(ctx, -1)) {
		/* Not an Error instance, don't read "stack". */
		return 1;
	}

	duk_get_prop_string(ctx, -1, "stack");  /* caller coerces */
	duk_remove(ctx, -2);
	return 1;
}

/* Print error to stderr and pop error. */
void
mn_print_pop_error(duk_context *ctx, FILE *f) {
	/* 
	 * Print error objects with a stack trace specially.
	 * Note that getting the stack trace may throw an error
	 * so this also needs to be safe call wrapped.
	 */
	(void)duk_safe_call(ctx, mn_get_stack_raw, 1 /*nargs*/, 1 /*nrets*/);
	fprintf(f, "%s\n", duk_safe_to_string(ctx, -1));
	fflush(f);
	duk_pop(ctx);
}

void *
mn_zmalloc(size_t size) {
	void *ret = calloc(1, size);
	return ret;
}

/* 
 * mn_stdin_to_tmpfile() is based on code from midipix. 
 * git.midipix.org/cgit.cgi/mdso/tree/src/driver/mdso_unit_ctx.c#n34
 * Huge thanks to midipix!
 */
FILE *
mn_stdin_to_tmpfile() {
	FILE *ftmp;
	char buf[4096];
	ssize_t	nread;
	int ret;

	if (!(ftmp = tmpfile())) {
		return 0;	
	}

	nread = read(0,buf,sizeof(buf)-1);

	while (nread) {
		if (nread > 0) {
			buf[nread] = '\0';
			ret = fputs(buf,ftmp);
		} else {
			ret = (errno == EINTR) ? 0 : -1;
		}

		if (ret < 0) {
			fclose(ftmp);
			return 0;
		}

		nread = read(0,buf,sizeof(buf)-1);
	}

	return ftmp;	
}

duk_ret_t
mn_mod_load(FILE *file, duk_context *ctx) {
	/*
	 * While the module loader will load files, it does
	 * so differently than the loader for the entry point.
	 * There will be no support for shebang lines here,
	 * it will just register objects in the context.
	 */
	return 1;
}

duk_ret_t
mn_mod_search(duk_context *ctx) {
	/*
	 * We get the following duk stack arguments:
	 *   index 0: id
	 *   index 1: require
	 *   index 2: exports
	 *   index 3: module
	 *
	 * For built-in modules, we need to map from, e.g., 'v8' to 
	 * dukopen_v8. This is done with a gperf(1) perfect hash,
	 * see src/include/builtin_hash.gperf for the details.
	 */
	const char *modname = duk_safe_to_string(ctx, 0);

	/*
	 * A required module prefixed with '/' is an absolute path.
	 * A required module prefixed with './' is a relative path.
	 *
	 * If the specified modname is a directory, look in it for
	 * a package.json file and include the 'main' entry there.
	 * 
	 * .js files are interpreted as JavaScript text files, and 
	 * .json files are parsed as JSON text files
	 * 
	 * NOTE: .node files are not supported.
	 */
	builtin *module = find_builtin(modname, strlen(modname));

	if (module) {
		module->loader(ctx);
		return 1;
	}

	/* Just punt on file lookup logic for now. */
	return 1;
}

int
main(int argc, char **argv) {

	duk_context *ctx = NULL; /* The heart of mininode! */
	int c = 0;               /* Used by getopt. */
	char *filename = NULL;   /* Kinda self-explanatory. */
	FILE *script = NULL;     /* Pointer to the script to execute */
	char *source = NULL;     /* The actual source to execute. */
	long srclen;             /* The length of source buffer */
	size_t srcsz;		 /* Size of NUL-terminated source buffer. */
	uv_loop_t *loop =  NULL; /* The event loop. */

	while ((c = getopt(argc, argv, OPTSTRING)) != -1) {
		switch (c) {
			case 'h':
				mn_short_help();
				exit(EXIT_SUCCESS);
			case 'v':
				printf("Version %s\n", MININODE_VERSION);
				exit(EXIT_SUCCESS);
			case 'p':
				if (print_flag == 0) {
					print_flag = 1;
				} else {
					goto sadplace;
				}
				break;
			case 'c':
				if (check_flag == 0) {
					check_flag = 1;
				} else {
					goto sadplace;
				}
				break;
			case 'i':
				if (interactive_flag == 0){
					interactive_flag = 1;
				} else {
					goto sadplace;
				}
				break;
			case 'z':
				if (zero_fill_flag == 0) {
					zero_fill_flag = 1;
				} else {
					goto sadplace;
				}
				break;
			default:
				goto sadplace;
		}
	}

	if (argc > optind) {
		/* Only accept one non-option file argument. */
		if (argc - optind > 1) {
			goto sadplace;
		} else {
			filename = argv[optind];
		}
	}

	/* Set the sigint handler and initialize the loop. */
	mn_set_sigint_handler();
	loop = uv_default_loop();

	/*
	 * See http://duktape.org/api.html#duk_create_heap
	 * 
	 * This might be an interesting place to use portableumem.
	 */
	if (zero_fill_flag) {
		fprintf(stderr, "WARNING: -z is currently unimplemented.\n");
		ctx = duk_create_heap(
			NULL, /* TODO: calloc() here */
			NULL,
			NULL,
			&loop,
			(void *)&mn_fatal_handler
		);
	} else {
		ctx = duk_create_heap(
			NULL,
			NULL,
			NULL,
			&loop,
			(void *)&mn_fatal_handler
		);
	}

	if (!ctx) {
		fprintf(stderr, "Failed to create a Duktape context!\n");
		exit(EXIT_FAILURE);
	}
	
	/* Detect piped input and use it as our script. */
	/* If we were invoked with -i, try to start the REPL. */
	if (!isatty(fileno(stdin)) && !interactive_flag) {
		filename = "/dev/stdin";
		script = mn_stdin_to_tmpfile();
	} else {
		/* If invoked without any file arguments, invoke the REPL. */
		if (argc == 1 || interactive_flag) {
			printf("repl is currently unimplemented.\n");
			goto sadplace;
		}

		/* If we can't access the file, exit. */
       		if (access(filename, R_OK) == -1) {
			fprintf(stderr, "Error: cannot read %s\n", filename);
       			goto sadplace;
       		} else {
       			script = fopen(filename, "r");
       		}
	}
	
	/*
	 * This loads the initial entry point either from the CLI or
	 * from a tmpfile created via stdin input. This entry script
	 * will have the *true* global __filename and __dirname.
	 */
	if (script != NULL) {
		if (fseek(script, 0L, SEEK_END) == 0) {
			srclen = ftell(script);

			if (srclen == -1) {
				fprintf(stderr, "An unknown error occurred!\n");
				exit(EXIT_FAILURE);
			}

			source = malloc(sizeof(char) * (srclen + 1));

			if (fseek(script, 0L, SEEK_SET) != 0) {
				fprintf(stderr, "An unknown error occurred!\n");
				exit(EXIT_FAILURE); 
			}
			
			/*
			 * TODO: Check for the shebang (#!/path/to/mininode)
			 * line here and remove it from the source buffer,
			 * so that files formatted as scripts may be ran
			 * when passed as file arguments or via stdin.
			 */

			srcsz = fread(source, sizeof(char), srclen, script);

			if (ferror(script) != 0) {
				fprintf(stderr, "Error reading file\n");
			} else {
				source[srcsz++] = '\0'; /* Just to be safe. */
			}
		}
		fclose(script);	
	} else {
		fprintf(stderr, "Could not open file!\n");
		exit(EXIT_FAILURE);
	}

	/*
	 * Register our module loader with the Duktape context.
	 */
	duk_get_global_string(ctx, "Duktape");
	duk_push_c_function(ctx, mn_mod_search, 4 /*nargs*/);
	duk_put_prop_string(ctx, -2, "modSearch");
	duk_pop(ctx);
	/*
	 * Register globals.
	 * See https://nodejs.org/dist/v6.2.2/docs/api/globals.html
	 *
	 * TODO: Actually implement all the builtin modules!
	 *
	 * Note that the 'timer' functions do not expose an
	 * object (viz 'console', 'process', 'buffer', etc). 
	 * The timer functions return an opaque object that 
	 * has the 'ref' and 'unref' methods.
	 *
	 * Also note that some built-ins may be invoked
	 * directly by other built-in functions without being
	 * included in the global namespace. This is useful
	 * with mn_bi_util_format() and mn_bi_console_log().
	 */
	mn_bi_errors(ctx);
	mn_bi_timers(ctx);
	mn_bi_events(ctx);
	mn_bi_buffer(ctx);
	mn_bi_process(ctx);
	mn_bi_console(ctx);

	/* TODO: Work out how to set __dirname and __filename! */
	/* See https://nodejs.org/dist/v6.2.2/docs/api/modules.html */

	/* Connect the Duk context with the event loop. */
	loop->data = ctx;

	if (!check_flag && !print_flag) {
		if (source != NULL) {
			/* 
			 * compile and run the contents of char *source
			 */
			if (duk_pcompile_lstring(ctx, 0, source, srclen) != 0) {
				printf(
					"compile failed: %s\n",
					duk_safe_to_string(ctx, -1)	
				);
			} else {
				free(source); /* Opportunistic! */
				duk_call(ctx, 0);
			}
			duk_pop(ctx);
			exit(EXIT_SUCCESS);
		} else {
			/* We should never get here. */
			fprintf(stderr, "An unknown error occurred!\n");
			exit(EXIT_FAILURE);
		}
	} else {
		fprintf(stderr, "-c and -p are currently unimplemented.\n");
		goto sadplace;
	}

	/* TODO: Support check_flag and print_flag */

	duk_destroy_heap(ctx);
	uv_loop_close(loop);
	exit(EXIT_SUCCESS);

	sadplace:
		mn_short_help();
		exit(EXIT_FAILURE);
}
